### DISCUSSION TITLE
Generic metforcing support

### DISCUSSION THREAD

**Discussion by pratiman-91** [NASA-LIS/LISF#830] [Ideas]
Creating custom met forcing is difficult as it required creating a new code and then compiling it into the LDT/LIS. I would like to propose a generic met forcing for generally used variables (shortwave, longwave, precip, temperature, pressure, specific humidity) format that can be directly used in the LDT/LIS. This could be as simple as the following:
```
netcdf dl{
dimensions:
	lon = 720 ;
	lat = 360 ;
	time = 2928 ;
variables:
	double lon(lon) ;
		lon:units = "degrees_east" ;
		lon:long_name = "Longitude" ;
	double lat(lat) ;
		lat:units = "degrees_north" ;
		lat:long_name = "Latitude" ;
	double time(time) ;
		time:long_name = "Time" ;
		time:units = "minutes since 1948-01-01 00:00" ;
	float dl(time, lat, lon) ;
		dl:_FillValue = -9.99e+08f ;
		dl:name = "dl" ;
		dl:long_name = "Downward longwave radiation" ;
		dl:units = "W m-2" ;
}
```
This could be extremely helpful, rather than writing, debugging, and then publishing code for each dataset. Please let me know what do you think about this. Thanks.

**Comment by dmocko** [comment]
Hi,

Thanks for your interest in LIS.

LIS does have a generic meteorological forcing option.  Please see the code in: LISF/lis/metforcing/genMetForc

There is an option to run LDT to generate this forcing, so in LIS, you would choose "LDT-generated" for the metforcing.

This generic metforcing reads in netCDF data.  You could skip the LDT step and generate the forcing outside of the LISF software.  You will need to ensure that the format of the netCDF data that you create is consistent with the genMetForc reader code.

@karsenau - you wrote this code in both LDT and LIS.  Do you have any further comments?

David

**Reply by pratiman-91** [reply]
Thank you @dmocko 

I saw the outputs of the test case NLDAS meteorological forcing. There are two precipitation variables ```Rainf_inst``` and ```CRainf_inst```. What is the difference between the two variables? Also, LDT process outputs for each timestep. It is possible to combine the timesteps. If yes, then what should be changed in the attributes. Thanks!

```
netcdf LDT_HIST_201006070300.d01 {
dimensions:
	east_west = 150 ;
	north_south = 100 ;
	time = 1 ;
variables:
	float lat(north_south, east_west) ;
		lat:units = "degree_north" ;
		lat:standard_name = "latitude" ;
		lat:scale_factor = 1.f ;
		lat:add_offset = 0.f ;
		lat:missing_value = -9999.f ;
		lat:_FillValue = -9999.f ;
		lat:vmin = 0.f ;
		lat:vmax = 0.f ;
...
```

**Reply by karsenau** [reply]
Hello @pratiman-91,

Thank you for your interest in LIS and for your suggestions.

Yes, as @dmocko pointed out, that LDT and LIS do support more "generic" netcdf-4 forcing readers, which you are welcome to use. We do provide information on how to use these forcing features in our "lis/configs/lis.config.adoc" file, found in our LISF repository.

As per your follow-up question, "CRainf_inst" reflects "convective rainfall" amounts, which some meteorological forcing datasets like NLDAS2 provide.  "Rainf_inst" reflects the amount of total rainfall which is the main input to any of our LSMs. Some models do utilize "CRainf_inst", if available.

These naming conventions follow the ALMA CF conventions, which you can learn more about here (like in Table 4.4):
https://www.umr-cnrm.fr/amma-moana/amma_surf/almip/docs/almip.pdf

We are also moving to even more NetCDF-based conventions, like the NetCDF CF Metadata Conventions, outlined here:
https://cfconventions.org

Hope this information helps.

We have also discussed supporting more "universal" format readers, like a netcdf-4 reader that support a variety of different netcdf datasets. However, that can take much time to cover all the different conventions.

**Reply by pratiman-91** [reply]
Thank you for the information, @karsenau.
Looking forward to the inclusion of NetCDF CF Metadata Conventions.

**Comment by hashmifatima** [comment]
Hi @karsenau 

I want use to NCUM Global data to use as metforcing in LIS. I have generated metforcing files like this: Tair_NCUM_202301.nc; Qair_NCUM_202301.nc etc.! Does this filename convention work? This file may contain data for the whole time period. Or should I have all the variable in one file?

Hashmi Fatima
NCMRWF, India

**Reply by dmocko** [reply]
Hi @hashmifatima 

Yes, that filename convention works.  You do not need to have all forcing variables in a single file.  You can design your reader to read them from separate files.

You can also design your reader to read data from files that contain the whole time period, contain a month's worth of data, or a single hour's worth (or any other way to organize in time).  However, you may find faster run performance in LIS if you break up your forcing files into time (such as one file - or set of files separated by variable name) based on your forcing data interval.  LIS will read in your forcing data quicker with smaller files instead of looking for the particular timestep in a large file.

**Reply by karsenau** [reply]
Hi @hashmifatima:

Thanks @dmocko for offering this very helpful information, and I would agree that you can create a metforcing reader to read in individual forcing files, or all forcing fields combined in one file. Either works.  Also, we have some readers that support forcing files that include multiple temporal points of data, or you can have separate files for each time point.  LIS is flexible and how you want to write that reader.

**Reply by hashmifatima** [reply]
Thank you very much @dmocko and @karsenau 

I have selected met forcing source as "LDT-generated" for NCUM metforcing in lis.config. It uses "genMetForc" plugin. There is a reader code inside genMetForc directory "readcrd_metForcGenerated.F90".  For this example I have 6-hourly data for 2 days. Every variable has separate data file.

**Do I need to write similar code for NCUM metforcing?** @karsenau

**Reply by hashmifatima** [reply]
Hi @karsenau 

There are these codes in the directory ~/LISF_557WW_release_7.4.2/lis/metforcing/genMetForc

0Intro_metForceGenerated.txt   
get_metForcGenerated.F90         
metForcGenerated_forcingMod.F90  
timeinterp_metForcGenerated.F90
finalize_metForcGenerated.F90  
metForcGen_SpatialInterpMod.F90  
readcrd_metForcGenerated.F90
get_metForcGen_filename.F90    
metForcGen_VariablesMod.F90      
reset_metForcGenerated.F90

Do I need to rewrite all of these codes for NCUM forcing or "metForcGenerated_forcingMod.F90" this only?

I am getting this error:
 [INFO] Generated metforcing directory: ./input/MET_FORCING/GENMETFORC
 [INFO] -- Obtain Forcing Dataset Parameters --
 [ERR] No LDT-generated forcing file,
 ./input/MET_FORCING/GENMETFORC/FORCING/202301/LDT_HIST_202301181100.d01.nc
 [ERR] is missing and can't extract file grid and time info.
 [ERR] Calling End Run ...
 [ERR] endrun is being called

Do I need to do some changes while running LDT also? 

Please clear this. 
Thank you 
Hashmi
